In [4]:
!pip install category_encoders

     |████████████████████████████████| 81kB 5.2MB/s 


# Sources


In [ ]:
#sources:
#https://www.kaggle.com/yasufuminakama/moa-pytorch-nn-starter
#https://www.kaggle.com/namanj27/new-baseline-pytorch-moa
#https://www.kaggle.com/headsortails/explorations-of-action-moa-eda

## Import libraries

In [5]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np

In [6]:
import os
import gc
import random
import math
import time
from sklearn.feature_selection import VarianceThreshold as V
import numpy as np
import pandas as pd

from sklearn.model_selection import StratifiedKFold, GroupKFold, KFold
from sklearn.metrics import log_loss

import category_encoders as ce

import torch
import torch.nn as nn
from torch.utils.data import DataLoader, Dataset
import torch.optim as optim
import torch.nn.functional as F

import warnings
warnings.filterwarnings("ignore")

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

def seed_everything(seed=42):
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.backends.cudnn.deterministic = True


/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


# Read The data

In [7]:
df_train = pd.read_csv('drive/MyDrive/lish-moa/train_features.csv')
df_test = pd.read_csv('drive/MyDrive/lish-moa/test_features.csv')
df_ss = pd.read_csv('drive/MyDrive/lish-moa/sample_submission.csv')
df_ts = pd.read_csv('drive/MyDrive/lish-moa/train_targets_scored.csv')
df_tns = pd.read_csv('drive/MyDrive/lish-moa/train_targets_nonscored.csv')

In [ ]:
print(f'shapes:\n train: {df_train.shape} \n target:{df_ts.shape} \n test: {df_test.shape}')

shapes:
 train: (23814, 876) 
 target:(23814, 207) 
 test: (3982, 876)


# Task
 <span style="color:#000000; font-family: 'Georgia'; font-size: 1.2em;">
Our task is multilabel classification. We are given 876 features(including id's of patients) and target consists of 206 features(each feature takes 0 or 1). <br>
df_train is our training set and df_ss is the target dataframe. Our task is to get target values for df_test. <br>
we additionally given df_tns - train_targets_nonscored. It is an additional (optional) binary responses for the training data. These are not predicted nor scored. I have no idea, why they gave it, so i didn't use it at all(as majority of competitors) <br>
link to the competition: https://www.kaggle.com/c/lish-moa/data
</span>

In [15]:
df_train.head(2)

sig_id cp_type  cp_time cp_dose  ...    c-96    c-97    c-98    c-99
0  id_000644bb2  trt_cp       24      D1  ... -0.3981  0.2139  0.3801  0.4176
1  id_000779bfc  trt_cp       72      D1  ...  0.1522  0.1241  0.6077  0.7371

[2 rows x 876 columns]

In [13]:
df_ts.head(2)

sig_id  ...  wnt_inhibitor
0  id_000644bb2  ...              0
1  id_000779bfc  ...              0

[2 rows x 207 columns]

In [14]:
df_test.head(2)

sig_id cp_type  cp_time cp_dose  ...    c-96    c-97   c-98   c-99
0  id_0004d9e33  trt_cp       24      D1  ...  0.0210 -0.0502  0.151 -0.775
1  id_001897cda  trt_cp       72      D1  ...  0.0359 -0.4764 -1.381 -0.730

[2 rows x 876 columns]

# QuanTile Transformation
 <span style="color:#000000; font-family: 'Georgia'; font-size: 1.2em;"> quatile transformation is one of the preprocssing steps, which makes the features to follow normal distribution. It also helps to reduce the effect of outliers</span>

In [ ]:
from sklearn.preprocessing import QuantileTransformer
num_cols = df_train.drop(columns = ['cp_time']).select_dtypes([np.number]).columns
qt = QuantileTransformer(n_quantiles=100, random_state=42, output_distribution = 'normal')

def quantileTransf(num_cols, df):
    for col in num_cols:
        qt.fit(df[col][:,None])
        df[col] = qt.transform(df[col][:,None])
    return df 

df_train = quantileTransf(num_cols, df_train)
df_test = quantileTransf(num_cols, df_test)    

# Drop ctrl_vehicle
 <span style="color:#000000; font-family: 'Georgia'; font-size: 1.2em;"> We drop the rows which has cp_type of ctl_vehicle, since their target values always 0, so we don't need them in the training, just in the end fill with 0 the rows with ctl_vehicle </span>

In [ ]:
def drop_ctlVecile(df):
    df = df[df['cp_type']!='ctl_vehicle'].reset_index(drop=True)
    return df

df_train = drop_ctlVecile(df_train)    
test = drop_ctlVecile(df_test)

data_with = pd.merge(df_train, df_ts, on='sig_id')
data = data_with[data_with['cp_type'] != 'ctl_vehicle']
print(f'New shapes: \n data(train(no ctl_vehicle) + target): {data.shape} \n test(no ctl_vehicle):                 {df_test.shape}')

New shapes: 
 data(train(no ctl_vehicle) + target): (21948, 1082) 
 test(no ctl_vehicle):                 (3982, 876)


# PCA - dimensionality reduction
 <span style="color:#000000; font-family: 'Georgia'; font-size: 1.2em;"> One might say that we don't need PCA, when you use Neural Network, since NN itself is good dimensionality reduction method. However, here the we concatenate values after PCA to the original values since it somehow improves the accuracy. This was noticed in one of the sources, i mentioned above </span>

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

gene = data.loc[:,'g-0': 'g-771']
gene_test = test.loc[:,'g-0': 'g-771']

scaler = StandardScaler()
scaled_gene = scaler.fit_transform(gene)
scaled_test_gene = scaler.transform(gene_test)
pca = PCA(0.95)
pca_data = pca.fit_transform(scaled_gene)
pca_test = pca.transform(scaled_test_gene)


In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

cell = data.loc[:,'c-0': 'c-99']
cell_test = test.loc[:,'c-0': 'c-99']

#--------------------------------------------------------------------------
scaler = StandardScaler()
scaled_cell = scaler.fit_transform(cell)
scaled_test_cell = scaler.transform(cell_test)

pca2 = PCA(0.95)
pca_data2 = pca2.fit_transform(scaled_cell)
pca_test2 = pca2.transform(scaled_test_cell)

In [ ]:
loadings = pca_data
labels = ['PC' + str(x) + '_g' for x in range(1, pca_data.shape[1] + 1)]
pca_gene = pd.DataFrame(loadings, columns = labels)
loadings2 = pca_data2
labels2 = ['PC' + str(x) + '_c' for x in range(1, pca_data2.shape[1] + 1)]
pca_cell = pd.DataFrame(loadings2, columns = labels2)
frames = [pca_gene, pca_cell]
df_gene_cell = pd.concat(frames, axis = 1)
print(f'So, the size of the new dataframe, which is concatenation of gene and cell features after PCA(0.95) is {df_gene_cell.shape}')

So, the size of the new dataframe, which is concatenation of gene and cell features after PCA(0.95) is (21948, 685)


In [ ]:
loadings = pca_test
labels = ['PC' + str(x) + '_g' for x in range(1, pca_test.shape[1] + 1)]
pca_test_gene = pd.DataFrame(loadings, columns = labels)
loadings2 = pca_test2
labels2 = ['PC' + str(x) + '_c' for x in range(1, pca_test2.shape[1] + 1)]
pca_test_cell = pd.DataFrame(loadings2, columns = labels2)
frames = [pca_test_gene, pca_test_cell]
df_pca_test = pd.concat(frames, axis = 1)
print(f'So, the size of the new dataframe, which is concatenation of gene and cell features after PCA(0.95) is {df_pca_test.shape}')

So, the size of the new dataframe, which is concatenation of gene and cell features after PCA(0.95) is (3624, 685)


In [ ]:
data = data.merge(df_gene_cell, left_index=True,right_index=True, how='left')
test = test.merge(df_pca_test, left_index=True,right_index=True, how='left')


# Neural Network

In [ ]:
import torch
import torch.nn as nn
import torch.nn.functional as F

In [ ]:
class base(nn.Module):
    def __init__(self):
        super(base, self).__init__()
        self.model = nn.Sequential(
            nn.BatchNorm1d(1560),
            nn.Dropout(0.2),
            nn.utils.weight_norm(nn.Linear(1560, 1024)),
            nn.BatchNorm1d(1024),
            nn.Dropout(0.5),
            nn.utils.weight_norm(nn.Linear(1024, 1024)),
            nn.BatchNorm1d(1024),
            nn.Dropout(0.5),
            nn.utils.weight_norm(nn.Linear(1024, 206)),
            
            
        
        )
       
    def forward(self, cont_x):
        x = self.model(cont_x)
        return x

## Train and test wrappers
 <span style="color:#000000; font-family: 'Georgia'; font-size: 1.2em;"> We have to wrap our data in order to be able to use them in DataLoader</span>

In [ ]:
class TrainDataset(Dataset):
    def __init__(self,df, numeric,  labels):
        self.numeric = df[numeric].values
        self.labels = labels
    def __len__(self):
        return len(self.numeric)
    def __getitem__(self, idx):
        numerical_tensor = torch.FloatTensor(self.numeric[idx])
        label = torch.tensor(self.labels[idx]).float()
        
        return numerical_tensor, label

    
class TestDataset(Dataset):
    def __init__(self,df, numeric):
        self.numeric = df[numeric].values
    def __len__(self):
        return len(self.numeric)
    def __getitem__(self, idx):
        numerical_tensor = torch.FloatTensor(self.numeric[idx])
        
        return numerical_tensor

    

# Variance Threshold
 <span style="color:#000000; font-family: 'Georgia'; font-size: 1.2em;"> I thought of this as another preprocessing method, but it greatly reduced the accuracy</span>


In [ ]:
'''
from sklearn.feature_selection import VarianceThreshold as V
selector = V(0.8)
my_cols = [i for i in test.columns if i not in ['sig_id', 'cp_type', 'cp_time', 'cp_dose']]
selected_data = selector.fit_transform(data.loc[:,my_cols])
selected_test = selector.transform(test.loc[:,my_cols])

In [ ]:
'''
labels = ['s' + str(x)  for x in range(1, selected_data.shape[1] + 1)]
selected_data = pd.DataFrame(selected_data, columns = labels)
selected_test = pd.DataFrame(selected_test, columns = labels)

target_cols = [i for i in df_ts.columns if i != 'sig_id']
categorical = ['cp_time', 'cp_dose']
target_cols += categorical
data = selected_data.merge(data[target_cols], left_index=True, right_index=True, how='left')
test = selected_test.merge(data[categorical], left_index=True, right_index=True, how='left')

# Categorical Variables
 <span style="color:#000000; font-family: 'Georgia'; font-size: 1.2em;">
we have some categorical values like cp_time, cp_dose, so i get their dummy values. I actually tried training without these categorical variables, but using their dummy variables improved the result </span>

In [ ]:
cp_time = pd.get_dummies(data['cp_time'], drop_first=True)
cp_dose = pd.get_dummies(data['cp_dose'], drop_first=True)
cp_time2 = pd.get_dummies(test['cp_time'], drop_first=True)
cp_dose2 = pd.get_dummies(test['cp_dose'], drop_first=True) 
test = test.merge(cp_time2, left_index=True,right_index=True, how='left')
test = test.merge(cp_dose2, left_index=True,right_index=True, how='left')
data = data.merge(cp_time, left_index=True,right_index=True, how='left')
data = data.merge(cp_dose, left_index=True,right_index=True, how='left')


numerical =  data.select_dtypes([np.number])
categorical = ['cp_time', 'cp_dose']
target_cols = df_ts.drop(columns = ['sig_id']).columns 
numerical = [i for i in numerical.columns if i not in categorical]
numerical = [i for i in numerical if i not in target_cols]
data['cp_time'] = data['cp_time'].map({24: 0, 48: 1, 72: 2}) 
data['cp_dose'] = data['cp_dose'].map({'D1': 3, 'D2': 4}) 
test['cp_time'] = test['cp_time'].map({24: 0, 48: 1, 72: 2}) 
test['cp_dose'] = test['cp_dose'].map({'D1': 3, 'D2': 4})

This ones simple training, validation and inference(for test set) functions \\
source: https://www.kaggle.com/yasufuminakama/moa-pytorch-nn-starter

In [ ]:
class AverageMeter(object):
    def __init__(self):
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count
        
def train(model, train_loader, scheduler, optimizer, device):
    losses = AverageMeter()

    model.train()
    for step, (numer, y) in enumerate(train_loader):
        
        numer, y = numer.to(device), y.to(device)
        batch_size = numer.size(0)

        pred = model(numer)
        loss = nn.BCEWithLogitsLoss()(pred, y)
        losses.update(loss.item(), batch_size)
       
    #    val_preds.append(pred.sigmoid().detach().cpu().numpy())

        
        loss.backward()
        
        grad_norm = torch.nn.utils.clip_grad_norm_(model.parameters(), 1000)

        scheduler.step()
        optimizer.step()
        optimizer.zero_grad()
        
    return losses.avg
        

def validation(model, valid_loader, device):
    losses = AverageMeter()
    model.eval()
    val_preds = []
    for step, (numer,  y) in enumerate(valid_loader):
        numer,  y = numer.to(device),  y.to(device)
        batch_size = numer.size(0)
        with torch.no_grad():
            pred = model(numer)
        loss = nn.BCEWithLogitsLoss()(pred, y)
        losses.update(loss.item(), batch_size)
        val_preds.append(pred.sigmoid().detach().cpu().numpy())
    val_preds = np.concatenate(val_preds)
    
    return losses.avg,val_preds        


def inference_fn(test_loader, model, device):

    model.eval()
    preds = []

    for step, (cont_x) in enumerate(test_loader):

        cont_x = cont_x.to(device)

        with torch.no_grad():
            pred = model(cont_x)

        preds.append(pred.sigmoid().detach().cpu().numpy())

    preds = np.concatenate(preds)

    return preds
       
            
   

# Loop
here we use multilableStratifiedKfold to divide our data into training and validation. This is the same as StratifiedKfold, but since our task is multilabel classification, we use multilabelStratifiedKFold

In [ ]:
#!pip install iterative-stratification

from iterstrat.ml_stratifiers import MultilabelStratifiedKFold
kf =  MultilabelStratifiedKFold(n_splits=5, random_state=42, shuffle=True)

kf.get_n_splits(data) 


train_cols = data.loc[:,"cp_time": ].columns
target_cols = df_ts.loc[:,"5-alpha_reductase_inhibitor": ].columns

def run_different_seed(data, target_cols, numerical, categorical, seed = 42):
    oof = np.zeros((len(data), len(target_cols)))
    predictions = np.zeros((len(test), len(target_cols)))

    seed_everything(seed = seed)
    
    print(f'Seed: {seed}')
    print('--------------------------------------')
    for fold_num, (train_index, test_index) in enumerate(kf.split(data, data[target_cols])):
        print(f'fold_num: {fold_num}')

        X_train, X_test = data.loc[:, train_cols], data.loc[:, train_cols]
        X_train, X_test = X_train.iloc[train_index], X_test.iloc[test_index]

        y_train, y_test = data[target_cols], data[target_cols]
        y_train, y_test = y_train.iloc[train_index].values, y_test.iloc[test_index].values

        train_dataset = TrainDataset(X_train, numerical, y_train)
        validation_dataset = TrainDataset(X_test, numerical,y_test)

        train_loader = DataLoader(train_dataset, batch_size = 32, shuffle=True, 
                                  num_workers=4, pin_memory=True, drop_last=True)
        valid_loader = DataLoader(validation_dataset, batch_size = 32, shuffle=False, 
                                  num_workers=4, pin_memory=True, drop_last=False)

        model = base()
        model.to(device)
        optimizer = optim.Adam(model.parameters(), lr=1e-2, weight_decay = 1e-6)
        scheduler = optim.lr_scheduler.OneCycleLR(optimizer=optimizer, pct_start=0.1, div_factor=1e3, 
                                                  max_lr=1e-2, epochs = 10, steps_per_epoch=len(train_loader))
        best_loss = np.inf
        for epochs in range(5):

            train_loss = train(model,train_loader, scheduler, optimizer, device)
            print(f'epoch: {epochs}')
            print(f'train: {train_loss}')
            valid_loss, val_preds = validation(model,valid_loader, device)
            print(f'validation: {valid_loss}')

            if valid_loss < best_loss:
                best_loss = valid_loss
                _oof = np.zeros((len(data), len(target_cols)))
                _oof[test_index] = val_preds
                print(f'saving the best model')
                torch.save(model.state_dict(), f"fold{fold_num}_seed{seed}.pth")

        test_dataset = TestDataset(test, numerical)#, categorical)

        test_loader = DataLoader(test_dataset, batch_size = 32, shuffle=False, 
                                  num_workers=4, pin_memory=True)
        model.load_state_dict(torch.load(f"fold{fold_num}_seed{seed}.pth"))
        model.to(device)
        print('------------------------------------------------------------------------')
        _predictions = inference_fn(test_loader, model, device)
        torch.cuda.empty_cache()
        
        oof += _oof
        predictions += _predictions / 5
        
    score = 0
    for i in range(data[target_cols].values.shape[1]):
        _score = log_loss(data[target_cols].values[:,i], oof[:,i])
        score += _score / data[target_cols].values.shape[1]
    print(f"CV score: {score}")
    print('-------------------------------------------------------------') 
    return oof, predictions

we use different seed for neural networks, it allows us to explore more weight spaces

In [ ]:
seeds = [1, 2, 3, 4, 5]
oof = np.zeros((len(data), len(target_cols)))
predictions = np.zeros((len(test), len(target_cols)))
for s in seeds:
    _oof, _predictions = run_different_seed(data,target_cols, numerical, categorical, seed = s)
    oof += _oof / len(seeds)
    predictions += _predictions / len(seeds)
score = 0
for i in range(data[target_cols].values.shape[1]):
    _score = log_loss(data[target_cols].values[:,i], oof[:,i])
    score += _score / data[target_cols].values.shape[1]
print(f"average CV score: {score}")

Seed: 1
--------------------------------------
fold_num: 0
epoch: 0
train: 0.20798605027690148
validation: 0.025065140463496668
saving the best model
epoch: 1
train: 0.020395829367798067
validation: 0.018458024897784468
saving the best model
epoch: 2
train: 0.01885358028344973
validation: 0.01785461714020283
saving the best model
epoch: 3
train: 0.017992535949324387
validation: 0.017570902978267376
saving the best model
epoch: 4
train: 0.017311564254078225
validation: 0.017312067488297514
saving the best model
------------------------------------------------------------------------
fold_num: 1
epoch: 0
train: 0.20848177641929283
validation: 0.021311235508472068
saving the best model
epoch: 1
train: 0.02078346256336646
validation: 0.0185871114916166
saving the best model
epoch: 2
train: 0.01894108926789006
validation: 0.01812619383792671
saving the best model
epoch: 3
train: 0.01807070003795254
validation: 0.01780620363364866
saving the best model
epoch: 4
train: 0.017407719497942793
va

In [ ]:
test[target_cols] = predictions
cols = [i for i in df_ts.columns]
cols = [i for i in df_ts.columns if i != 'sig_id']
sub = df_ss.drop(columns=[i for i in cols]).merge(test[cols], left_index=True,right_index=True, how='left').fillna(0)
sub.to_csv('submission.csv', index=False)
sub.head()

sig_id  ...  wnt_inhibitor
0  id_0004d9e33  ...       0.001354
1  id_001897cda  ...       0.001103
2  id_002429b5b  ...       0.001132
3  id_00276f245  ...       0.000951
4  id_0027f1083  ...       0.001263

[5 rows x 207 columns]